src/toolchain.ml

```diff
@@ -184,6 +184,63 @@ let of_settings () : t option Promise.t =
   | Some Global -> Promise.return (Some Global)
   | Some (Custom template) -> Promise.return (Some (Custom template))
 
+let detect_esy_sandbox ~project_root esy () =
+  let open Promise.Syntax in
+  let* esy = esy in
+  let* esy_build_dir_exists =
+    Fs.exists Path.(project_root / "_esy" |> Path.to_string)
+  in
+  let+ manifest = Esy.find_manifest_in_dir project_root in
+  match (esy, esy_build_dir_exists, manifest) with
+  | Some esy, true, _
+  | Some esy, _, Some _ ->
+    (* Esy can be used with [esy.json], [package.json], or without any of those.
+        So we check if we find an [_esy] directory, which means the user created an Esy sandbox.
+
+       If we don't, but there is an [esy.json] file, we can assume the user wants to use Esy.
+    *)
+    Some (Esy (esy, project_root))
+  | _ -> None
```

Comment:
By implementing the previous comment, we can avoid this large pattern match & wildcard in it.